### ISSUE TITLE
How can I build OSMDroid with Gradle on Android Arctic Fox IDE without using Gradle-Fury?

### ISSUE THREAD

**Issue by eclectice** [osmdroid/osmdroid#1756]
Please note: issues related to or caused by osmbonuspack will be closed with a
suggestion to open an issue at https://github.com/MKergall/osmbonuspack

(feel free to delete whatever doesn't apply)

## Issue Type

[ ] Question
[ ] Bug
[ ] Improvement
[X ] Build system related
[ ] Performance
[ ] Documentation

## Description and/or steps/code to reproduce the problem

Please provide an alternative approach to build the OSMDroid source codes using Gradle in Android Studio Arctic Fox that does not require Gradle-Fury.

Currently, what I have done so far:
1. remove generated folders in the project root since I need to upgrade Gradle plugin to 7.0.3 and Gradle wrapper dist to 7.0.2:
```
.idea
.gradle
```
2. Upgrade `distributionUrl` in the gradle-wrapper.properties
`distributionUrl=https\://services.gradle.org/distributions/gradle-7.0.2-bin.zip`

2. comment out any mention of `apply from: "https://raw.githubusercontent.com/gradle-fury/gradle-fury/v1.1.4/gradle/{whatever}.gradle"` in every build.gradle file
3. Explicitly enforce the following codes encountered:
```
def isAndroidApplication = true
def canSign = false

afterEvaluate { project ->
    tasks.each { task ->
        if (isAndroidApplication) {
            if ((task.name.toLowerCase().startsWith("assemble") && task.name.toLowerCase().contains('release'))
                    || (task.name.toLowerCase().startsWith("bundle") && task.name.toLowerCase().contains('release'))
                    || (task.name.toLowerCase().startsWith("package") && task.name.toLowerCase().contains('release'))
                    || (task.name.equals("validateSigningRelease"))) {

                task.setEnabled(canSign)
            }
        }
    }
}

```
4. Change deprecated Android Support libraries to equivalent AndroidX libraries (still trying to resolve so many issues)

5. Explicitly include the required attributes for the OpenStreetMapViewer Android application:
```
    compileSdkVersion 31
    buildToolsVersion "30.0.3"

```
6. Add into the root project's gradle.properties file:
```
android.useAndroidX=true
android.enableJetifier=true
```
7. Use hard-coded version values for most of the package dependencies
8. Remove `apply plugin: maven` since I just need to create the apps locally
9. Change to new Gradle syntaxes in the dependencies.
10. Changed application manifest `allowBackup` attribute since all apps are just for testing/demo:
```
    <application
        android:allowBackup="false" tools:replace="android:allowBackup"
        android:fullBackupOnly="false"
        
```
11. Change the buildscript entry in the project's build.gradle into:
```
buildscript {
    repositories {
        mavenCentral()
        mavenLocal()
        maven {
            url 'https://maven.google.com/'
            name 'Google'
        }
    }
    dependencies {
        classpath "com.android.tools.build:gradle:7.0.3"
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.31"
    }
}
```
12. Change the JVM memory size in gradle.properties
`org.gradle.jvmargs=-Xms1536m -Xmx2048m -XX:MaxPermSize=256m`
13. Add `android:exported="true"` in all activity entries in the application manifest
14. Add VERSION_NAME and BUILD_TYPE for the BuildConfig in the OSMStreetMapViewer's build.gradle
```
    defaultConfig {
        applicationId 'org.osmdroid'
        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"

        minSdkVersion Integer.parseInt(project.property('android-minSdkForSupportLib.version'))
        multiDexEnabled = true

        buildConfigField("String", "VERSION_NAME", "\"${project.version}\"")
    }

    buildTypes {
        release {
            buildConfigField("String", "BUILD_TYPE", "\"release\"")
        }
        debug {
            buildConfigField("String", "BUILD_TYPE", "\"debug\"")
        }
    }

    compileOptions {
        coreLibraryDesugaringEnabled true

        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
```
## Environment

```
Android Studio Arctic Fox | 2020.3.1 Patch 3
Build #AI-203.7717.56.2031.7784292, built on October 1, 2021
Runtime version: 11.0.10+0-b96-7249189 amd64
VM: OpenJDK 64-Bit Server VM by Oracle Corporation
Windows 10 10.0
GC: G1 Young Generation, G1 Old Generation
Memory: 1280M
Cores: 4
Registry: analyze.exceptions.on.the.fly=true, external.system.auto.import.disabled=true, ide.images.show.chessboard=true
Non-Bundled Plugins: com.alayouni.ansiHighlight, com.intellij.ideolog, JFormDesigner, idea.plugin.protoeditor, io.protostuff.protostuff-jetbrains-plugin, Dart, org.jetbrains.kotlin, io.flutter, org.intellij.plugins.markdown
```

### If it's a bug, version(s) of android this affects:

All

### Version of osmdroid the issue relates to:

6.1.11

**Comment by eclectice** [comment]
I forked this project and made the following changes in my fork: https://github.com/eclectice/osmdroid

Another notable change that I forgot to mention in my push message is that I have added the Context parameter to most of the OSMDroid classes to deal with the deprecated BitmapDrawable(Bitmap) when trying to understand the problem of rendering bitmaps from the data blobs of the GeoPackager raster tiles.

**Comment by LostRat** [comment]
Look at my fork. (lostrat/osmdroid) This took a while for me to figure out.
Never could figure out maven local. Still trying. Still trying to figure
out JitPak and mavenlocal.

Clone this locally:

https://github.com/LostRat/osmdroid/tree/migrate-to-androidx

use terminal, after you navigate to a parent folder
git clone [use link you get from github]

Then open folder from Android Studio.

Here is the trick I only figured out a couple days ago. MavenLocal is not
mandatory for local build from your library. In your app that you want to
use the library edit the project settings.gradle file to add a line like
this:

includeBuild  'F:/ANDROIDPROJECT/osmdroid'

Invoke Gradle sync, and then marvel at what the IDE developers developed.

You might use Meld to compare your fork at home with mine that you cloned.
I tried to do my commits and branches in a sequence for someone to figure
out and repeat on their own. I can build and run the sample app, and now my
own app.
